homeassistant/components/mysensors/switch.py

```diff
@@ -32,43 +47,56 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
         "S_MOISTURE": MySensorsSwitch,
         "S_WATER_QUALITY": MySensorsSwitch,
     }
-    mysensors.setup_mysensors_platform(
-        hass,
-        DOMAIN,
-        discovery_info,
-        device_class_map,
-        async_add_entities=async_add_entities,
-    )
 
-    async def async_send_ir_code_service(service):
-        """Set IR code as device state attribute."""
-        entity_ids = service.data.get(ATTR_ENTITY_ID)
-        ir_code = service.data.get(ATTR_IR_CODE)
-        devices = mysensors.get_mysensors_devices(hass, DOMAIN)
-
-        if entity_ids:
-            _devices = [
-                device
-                for device in devices.values()
-                if isinstance(device, MySensorsIRSwitch)
-                and device.entity_id in entity_ids
-            ]
-        else:
-            _devices = [
-                device
-                for device in devices.values()
-                if isinstance(device, MySensorsIRSwitch)
-            ]
-
-        kwargs = {ATTR_IR_CODE: ir_code}
-        for device in _devices:
-            await device.async_turn_on(**kwargs)
-
-    hass.services.async_register(
-        MYSENSORS_DOMAIN,
-        SERVICE_SEND_IR_CODE,
-        async_send_ir_code_service,
-        schema=SEND_IR_CODE_SERVICE_SCHEMA,
+    async def async_discover(discovery_info):
+        """Discover and add a MySensors switch."""
+        mysensors.setup_mysensors_platform(
+            hass,
+            DOMAIN,
+            discovery_info,
+            device_class_map,
+            async_add_entities=async_add_entities,
+        )
+
+        async def async_send_ir_code_service(service):
+            """Set IR code as device state attribute."""
+            entity_ids = service.data.get(ATTR_ENTITY_ID)
+            ir_code = service.data.get(ATTR_IR_CODE)
+            devices = mysensors.get_mysensors_devices(hass, DOMAIN)
+
+            if entity_ids:
+                _devices = [
+                    device
+                    for device in devices.values()
+                    if isinstance(device, MySensorsIRSwitch)
+                    and device.entity_id in entity_ids
+                ]
+            else:
+                _devices = [
+                    device
+                    for device in devices.values()
+                    if isinstance(device, MySensorsIRSwitch)
+                ]
+
+            kwargs = {ATTR_IR_CODE: ir_code}
+            for device in _devices:
+                await device.async_turn_on(**kwargs)
+
+        hass.services.async_register(
```

Comment:
In a future PR we should change to use the entity service helper:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

Reply:
We should only register the service once, in the scope of `async_setup_entry`.